### PR TITLE
Rewrite emdump: JS+Wasm size reporting tool

### PR DESCRIPTION
**Pull request by juj**
This PR deletes the old `tools/emdump` tool, and rewrites it under `tools/size_report.js`. Emdump was not able to process wasm sizes, but `size_report.js` is able to process both JS and Wasm.

An important feature of this tool is to be able to work when Closure is being used, and/or when `--emit-symbol-map` is being used (as opposed to just `--profiling-funcs` or `-g2` is being used)

Current output from the tool looks like:
```
$ node tools/size_report.js cpuprofiler.wasm cpuprofiler.js --symbols cpuprofiler.symbols --expandLargerThanBytes 1000
--- Code sizes:
cpuprofiler.wasm/function dlmalloc: 5985
cpuprofiler.wasm/section DATA: 3201
cpuprofiler.wasm/function __rem_pio2_large: 1713
cpuprofiler.js/var asmLibraryArg: 1006
cpuprofiler.js/function _emscripten_webgl_do_create_context: 974
cpuprofiler.wasm/function draw_frame(double, void*): 839
cpuprofiler.js/var GL.initExtensions/automaticallyEnabledExtensions: 793
cpuprofiler.js/function GL.populateUniformTable: 740
cpuprofiler.js/function JSEvents.registerOrRemoveHandler: 721
cpuprofiler.js/function _glUniformMatrix4fv: 705
cpuprofiler.js/function UTF8ArrayToString: 677
cpuprofiler.js/function _upload_unicode_char_to_texture: 662
cpuprofiler.js/function JSEvents.deferCall: 549
cpuprofiler.wasm/function memcpy: 523
cpuprofiler.js/function __webgl_enable_ANGLE_instanced_arrays: 488
cpuprofiler.wasm/function fill_image: 485
cpuprofiler.wasm/function init_webgl: 481
cpuprofiler.wasm/function fmodf: 463
cpuprofiler.js/function GL.initExtensions: 453
cpuprofiler.js/function GL.registerContext: 443
cpuprofiler.js/function _glGetUniformLocation: 437
cpuprofiler.wasm/function fill_char: 435
cpuprofiler.js/function __webgl_enable_OES_vertex_array_object: 411
cpuprofiler.js/function emscriptenWebGLGetTexPixelData: 407
cpuprofiler.js/var JSEvents: 402
cpuprofiler.wasm/function sinf: 383
cpuprofiler.js/function GL.deleteContext: 382
cpuprofiler.js/function GL.init: 365
cpuprofiler.js/var GL: 361
cpuprofiler.js/function __computeUnpackAlignedImageSize: 264
cpuprofiler.js/function JSEvents.removeAllHandlersOnTarget: 264
cpuprofiler.js/function _glTexImage2D: 260
cpuprofiler.wasm/section IMPORT: 259
cpuprofiler.js/function __glGenObject: 259
cpuprofiler.wasm/function __rem_pio2f: 259
cpuprofiler.js/function JSEvents.runDeferredCalls: 252
...
```

although I am thinking of restylizing this to similar format of emdump before landing.

Btw, it was originally proposed that `emdump` should reside in the root directory, but that [was rejected](https://github.com/emscripten-core/emscripten/pull/6967#issuecomment-411823457) citing that the tool was not useful enough, and not tested. Now I see that there is a utility `emsize` that was added in #9663, which is quite objectively even less useful than `emdump`. I think the review process has not been very consistent here.

The reason I am raising that up is that there is a big conveniency/ux aspect to having a tool reside in root directory of Emscripten vs in a subdirectory: if the tool is placed in the root directory, one can invoke it easily with

```bash
/home/clb/myproject$ emdump myfile.js
```

when one has Emscripten in PATH, e.g. by having an active EMSDK command prompt set.

Whereas in a tools subdirectory, one will need to type

```bash
/home/clb/myproject$ python $EMSCRIPTEN/tools/emdump.py myfile.js
```
(or `$EMSCRIPTEN/tools/emdump myfile.js` depending on whether exec stubs are present) to run the tool, which is much more inconvenient.

I would argue the benefit of `emsize` is a bit vague in comparison with this new `size_report.js` tool. I would be happy to place `size_report.js` to the root directory under e.g. the name `emdump`, or replace `emsize` with this.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -3362,12 +3362,14 @@ def preprocess_wasm2js_script():
 
   def run_closure_compiler(final):
     final = building.closure_compiler(final, pretty=not optimizer.minify_whitespace,
```
Comment (sbc100) [inline emcc.py:3364]: Can you split these args onto newlines?

File: tools/building.py [hunk 2]
```diff
@@ -1153,11 +1157,9 @@ def add_to_path(dirname):
 
     proc = run_process(cmd, stderr=PIPE, check=False, env=env)
 
-    # XXX Closure bug: if Closure is invoked with --create_source_map, Closure should create a
+    # XXX Closure bug: Closure generates a stray output file even when not requested, so delete
     # outfile.map source map file (https://github.com/google/closure-compiler/wiki/Source-Maps)
-    # But it looks like it creates such files on Linux(?) even without setting that command line
-    # flag (and currently we don't), so delete the produced source map file to not leak files in
-    # temp directory.
+    # the produced source map file to not leak files in temp directory.
```
Comment (sbc100) [inline tools/building.py:1162]: Does this happen if if you pass `--create_source_map` above?

File: tools/building.py [hunk 3]
```diff
@@ -1197,6 +1199,13 @@ def add_to_path(dirname):
       if Settings.CLOSURE_WARNINGS == 'error':
         exit_with_error('closure compiler produced warnings and -s CLOSURE_WARNINGS=error enabled')
 
+      # Convert the source map to a symbol map if we are generating symbol maps.
+      if emit_symbol_map:
+        symbol_map_data = subprocess.check_output(NODE_JS + [path_from_root('tools', 'size_report.js'), '--createSymbolMapFromSourceMap', source_map, outfile])
```
Comment (sbc100) [inline tools/building.py:1204]: Is this logically be a separate conversion tool?convertion

File: tools/size_report.js [hunk 4]
```diff
@@ -0,0 +1,1077 @@
+var acorn = require('../node_modules/acorn');
```
Comment (sbc100) [inline tools/size_report.js:1]: Copyright + top level overview comment

**Comment by sbc100** [comment]
If we are going to start writing more of the compiler driver in JS then I think a few things need to change.

Firstly I need to learn more JS so I can fully review it :) 

But also we are going to want linters and style checkers for this JS.  We to take advantage of modern JS features and conventions.  We are going to want to take advantage of the latest node module, etc.   I'm not against it I just think there are a lot of things we (perhaps mostly I) are currently missing to be able to deal with a lot of new JS code in the driver.

Personally I'd rather stick with python for now until we come to shared decision, but obviously asking your re-write all this code is a lot to ask.

**Comment by Akaricchi** [comment]
> This is problem right now with people who are trying to use file_packager.py directly (I'm not sure we want to encourage that).

The problem with `file_packager.py` is that it's impossible to do some useful things *without* calling it directly. This is the only way to split assets into multiple packages. It's also useful to separate packing from the rest of the build so that the build system can cache the build artifacts and better parallelize the jobs. In my use case the packages are rebuilt much less frequently than the code is, so that helps a lot. I think it makes a lot of sense to promote `file_packager.py` to an official user-facing tool and put it in `PATH` (call it `empack` maybe), rather than trying to discourage its use. I also agree that this new tool would be nice to have in `PATH` as well.

**Comment by sbc100** [comment]
> > This is problem right now with people who are trying to use file_packager.py directly (I'm not sure we want to encourage that).
> 
> The problem with `file_packager.py` is that it's impossible to do some useful things _without_ calling it directly. This is the only way to split assets into multiple packages. It's also useful to separate packing from the rest of the build so that the build system can cache the build artifacts and better parallelize the jobs. In my use case the packages are rebuilt much less frequently than the code is, so that helps a lot. I think it makes a lot of sense to promote `file_packager.py` to an official user-facing tool and put it in `PATH` (call it `empack` maybe), rather than trying to discourage its use. I also agree that this new tool would be nice to have in `PATH` as well.

Thats fine with me.   I agree we should probably rename to have an `em` prefix.

**Pull request closed without merging**